framework/Models/HybridModel.py

```diff
@@ -0,0 +1,621 @@
+          romInput.append(elem)
+        else:
+          self.raiseAnError(IOError, "The type of input ", elem.name, " can not be accepted!")
+      if self.romValid:
+        return (romInput, samplerType, newKwargs)
+      else:
+        return (codeInput, samplerType, newKwargs)
+
+    return (myInput, samplerType, newKwargs)
+
+  def trainRom(self, samplerType, kwargs):
+    """
+      This function will train all ROMs if they are not converged
+      @ In, None
+      @ Out, None
+    """
+    self.raiseADebug("Start to train roms")
+    self.raiseADebug("Current sample: ", self.counter)
+    for romInfo in self.romsDictionary.values():
+      # reset the rom
+      romInfo['Instance'].amITrained = False
+      # always train the rom even if the rom is converged, we assume the cross validation and rom train are relative cheap
+      outputMetrics = self.cvInstance.evaluateSample([romInfo['Instance'], self.tempTargetEvaluation], samplerType, kwargs)[1]
+      converged = self.isRomConverged(outputMetrics)
+      romInfo['Converged'] = converged
+      if converged:
+        romInfo['Instance'].train(self.tempTargetEvaluation)
+        self.raiseADebug("ROM ", romInfo['Instance'].name, " is converged!")
+
+  def isRomConverged(self, outputDict):
+    """
+      This function will check the convergence of rom
+      @ In, outputDict, dict, dictionary contains the metric information
+      @ Out, None
+    """
+    converged = True
+    for targetName, metricInfo in outputDict.items():
+      for metricName, metricValues in metricInfo.items():
+        if max(metricValues) > self.romConvergence:
+          self.raiseADebug("maximum error: ", max(metricValues), " is greater than the given tolerance")
+          converged = False
+    return converged
+
+  def checkRomConvergence(self):
+    """
+      This function will check the convergence of all ROMs
+      @ In, None
+      @ Out, bool, True if all ROMs are converged
+    """
+    converged = True
+    for romInfo in self.romsDictionary.values():
+      if not romInfo['Converged']:
+        converged = False
+    if converged:
+      self.raiseADebug("All ROMs are converged")
+    return converged
+
+  def checkRomValidity(self, kwargs):
+    """
+      This function will check the validity of all roms
+      @ In, kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+        a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, None
+    """
+    allValid = False
+    if self.modelSelection == 'CrowdingDistance':
+      allValid = self.crowdingDistanceMethod(kwargs)
+    else:
+      self.raiseAnError(IOError, "Unknown model selection method ", self.modelSelection, " is given!")
+
+    if allValid:
+      self.raiseADebug("ROMs  are all valid for given model ", self.modelInstance.name)
+
+    return allValid
+
+  def crowdingDistanceMethod(self, kwargs):
+    """
```

Comment:
why you pass the all kwargs if you just need the SampledVars?

Reply:
fixed